gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java

```diff
@@ -364,6 +369,112 @@ public void testOverwrite()
     // Try to create the same file again with overwrite == false.
     assertThrows(
         IOException.class, () -> ghfsHelper.writeFile(hadoopPath, text, 1, /* overwrite= */ false));
+
+    // Try to create the same file again with overwrite == true.
+    ghfsHelper.writeFile(hadoopPath, text, 1, /* overwrite= */ true);
+    String readText = ghfsHelper.readTextFile(hadoopPath, 0, text.getBytes().length, true);
+    assertThat(readText).isEqualTo(text);
+  }
+
+  // TODO: Re-enable this for unit test after updating local ImMemoryGoogleCloudStorage to mirror
+  // prod GCS behavior for concurrent overwrite.
+  @Test
+  @Ignore
+  public void testConcurrentCreationWithoutOverwrite_onlyOneSucceeds() throws IOException {
+    // Get a temp path and ensure that it does not already exist.
+    URI path = GoogleCloudStorageFileSystemIntegrationTest.getTempFilePath();
+    Path hadoopPath = ghfsHelper.castAsHadoopPath(path);
+    assertThrows(FileNotFoundException.class, () -> ghfs.getFileStatus(hadoopPath));
+
+    String text = "Hello World!";
+    ExecutorService executorService = Executors.newFixedThreadPool(2);
+
+    assertThrows(
+        ExecutionException.class,
+        () -> {
+          Future<Integer> future1 =
+              executorService.submit(
+                  () -> ghfsHelper.writeFile(hadoopPath, text, 1, /* overwrite= */ false));
+          Future<Integer> future2 =
+              executorService.submit(
+                  () -> ghfsHelper.writeFile(hadoopPath, text, 1, /* overwrite= */ false));
+          // wait until result will be ready
+          Integer numBytesWritten1 = future1.get();
+          Integer numBytesWritten2 = future2.get();
+          assertThat(numBytesWritten1 + numBytesWritten2).isEqualTo(text.getBytes(UTF_8).length);
+        });
+
+    executorService.shutdown();
+
+    // Verify at least one creation request succeeded.
+    String readText = ghfsHelper.readTextFile(hadoopPath, 0, text.getBytes().length, true);
+    assertThat(readText).isEqualTo(text);
+
+    // Verify it can be overwritten by another creation.
+    Integer numBytesWritten = ghfsHelper.writeFile(hadoopPath, text, 1, /* overwrite= */ true);
+    assertThat(numBytesWritten).isEqualTo(text.getBytes(UTF_8).length);
+  }
+
+  @Test
+  public void testConcurrentCreationWithOverwrite_bothSucceed()
+      throws IOException, ExecutionException, InterruptedException {
+    // Get a temp path and ensure that it does not already exist.
+    URI path = GoogleCloudStorageFileSystemIntegrationTest.getTempFilePath();
+    Path hadoopPath = ghfsHelper.castAsHadoopPath(path);
+    assertThrows(FileNotFoundException.class, () -> ghfs.getFileStatus(hadoopPath));
+
+    String text = "Hello World!";
+    ExecutorService executorService = Executors.newFixedThreadPool(2);
+
+    Future<Integer> future1 =
+        executorService.submit(
+            () -> ghfsHelper.writeFile(hadoopPath, text, 1, /* overwrite= */ true));
```

Comment:
Write different content and validate that file content equals to either of them.

Reply:
Done.